### PR TITLE
Change Payments menu and tasks behavior depending on React Payments page feature flag

### LINKED ISSUE #53379: [WC Payments NOX] Redirect from get paid task and payments menu item to the settings page rather than wcpay welcome page
### Description
when WooPayments is not active:
- Redirect from payments menu item to the Payments settings page rather than payments task page
- Redirect from get paid task to the Payments settings page.
- Redirect from the Set up additional payment options task to the Payments settings page.

Image

### Dev notes
Only if the feature flag is on!

### Designs
Design post: pfHfG4-q7-p2
Figma: xO8YdpKfaVoP2d8NvweZ5I-fi-1428_21437

### Dev notes
Reactifying the Payments settings page: pe4Hk0-1ly-p2
Populate react-powered main payments settings screen PR: https://github.com/woocommerce/woocommerce/pull/50825

### Additional context
PRD: 1wMfX8CAjDnOp24tcU-PADB0pOt3cz7mGdyZYeRBxx2Y-gdoc

### PR DESCRIPTION
**Pull request by vladolaru**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

If the `reactify-classic-payments-settings` feature flag is enabled, we link from the Payments menu item, the "Get paid" and "Set up additional payment options" tasks to the new Payments settings page.
We deactivated the WooPayments-specific task (the one with the title "Get paid with WooPayments") and the incentives page (WCPay Welcome Page).

Closes #53379 .

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Pre-testing setup

1. Create a new JurassicNinja site with WooCommerce on this PR's live branch and the WooCommerce Beta Tester plugin (here is a [direct JN create link](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce-beta-tester&woocommerce&branches.woocommerce=update/53379-link-payments-menu-and-task-to-nox))
2. Go to WP admin > Tools > WCA Test Helper > Features and make sure the `reactify-classic-payments-settings` is NOT enabled

#### Test the current behavior

1. Go through [this PR's testing instructions](https://github.com/woocommerce/woocommerce/pull/51541), minus the first one to set up the JN site - you already did that.
2. You should now have WooPayments set up.
3. Go to WooCommerce > Home and hide the WooCommerce setup task list:
Screenshot 2024-11-15 at 20 22 36
4. Go to Plugins and deactivate the WooPayments plugin
5. Click on the "Payments" top-level main menu item and will be taken to the Payments settings page:
Screenshot 2024-11-15 at 20 29 09
6. Go to Plugins and activate the "WooCommerce Stripe Payment Gateway" (it is already set up with an account)
7. Click on the "Payments" top-level main menu item and you will be taken to the "Set up additional payment options" task page:
Screenshot 2024-11-15 at 20 36 51
8. Go to WooCommerce > Home and hide the "Things to do next" task list:
Screenshot 2024-11-15 at 20 38 16
9. Refresh the page, click on the "Payments" top-level main menu item, and you will be taken to the Payments settings page.
10. Activate back WooPayments and when you click on "Payments" top-level main menu item you will be taken to the WooPayments overview page:
Screenshot 2024-11-15 at 20 41 36

#### Test with the feature flag enabled

1. Deactivate WooPayments and the WooCommerce Stripe Payment Gateway plugins from the previous steps.
2. Go to WooCommerce > Settings > Payments and make sure no offline payments methods are not enabled.
3. Go to WP admin > Tools > WCA Test Helper > Features and make enable the `reactify-classic-payments-settings` feature flag.
4. Go to WP admin > Tools > WCA Test Helper > Options, search for `woocommerce_task_list_hidden_lists` and Delete the option to make the task lists reappear.
5. Go to WooCommerce > Home, and you should see a "Get paid" task. Click on it and you will be taken to the new Payments settings page:
Screenshot 2024-12-06 at 16 08 59
6. Go to WooCommerce > Home, click on the Payments menu item (with a red badge notice), and you will be taken to the new Payments settings page.
7. Click on Take offline payments and enable "Cash on delivery"
8. Go to WooCommerce > Home, and the "Get paid" task should be marked as completed, the Payments main menu item should not have a red badge notice anymore, and a "Set up additional payment options" task should be visible in the Things to do next list:
Screenshot 2024-12-06 at 16 14 03
9. Click on the "Set up additional payment options" task and you will be taken to the new Payments settings page with the "Other payment options" section expanded.
10. Go to WooCommerce > Home, and hide the main task list.
11. Refresh the page, click on the Payments menu item, and you should be taken to the new Payments settings page.
12. Go to WooCommerce > Home, and hide the "Things to do next" task list.
13. Refresh the page, click on the Payments menu item, and you should be taken to the new Payments settings page.
14. Go to WP admin > Tools > WCA Test Helper > Options, search for `wcpay_welcome_page_incentives_dismissed`, and Delete the option.
15. Refresh the page, click on the Payments menu item, and you should be taken to the new Payments settings page.
16. Enable WooPayments:
Screenshot 2024-12-06 at 16 22 28
17. Refresh the page, click on the Payments menu item (that now has sub-menu items), and you should be taken to the WooPayments overview page.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Internal/Admin/Settings/PaymentsController.php [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+use Automattic\WooCommerce\Admin\Features\Features;
+use Exception;
+
+defined( 'ABSPATH' ) || exit;
+/**
+ * Payments settings controller class.
+ *
+ * Use this class for hooks and actions related to the Payments settings page.
+ */
+class PaymentsController {
```
Comment (vladolaru) [inline plugins/woocommerce/src/Internal/Admin/Settings/PaymentsController.php:15]: I don't see much benefit from introducing unit tests for the methods here (at least not at this point). We should be pretty covered with manual testing.

File: plugins/woocommerce/src/Admin/Features/OnboardingTasks/Tasks/Payments.php [hunk 2]
```diff
@@ -97,4 +102,19 @@ function( $gateway ) {
...
+	 * @return string
+	 */
+	public function get_action_url() {
+		// If the React-based Payments settings page is enabled, we want the task to link to the Payments Settings page.
+		if ( Features::is_enabled( 'reactify-classic-payments-settings' ) ) {
+			return admin_url( 'admin.php?page=wc-settings&tab=checkout' );
+		}
+
+		// Otherwise, we want the task behavior to remain unchanged (link to the Payments task page).
+		return '';
```
Comment (oaratovskyi) [inline plugins/woocommerce/src/Admin/Features/OnboardingTasks/Tasks/Payments.php:118]: a bit weird code-wise - returning empty string will remain the behaviour unchanged, but it works 🤷

File: plugins/woocommerce/src/Internal/Admin/Settings/PaymentsController.php (hunk 1 above)
Comment (oaratovskyi) [reply plugins/woocommerce/src/Internal/Admin/Settings/PaymentsController.php:15]: sounds good to me. It was tested except needs setup use case.

File: plugins/woocommerce/src/Internal/Admin/Settings/PaymentsController.php [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+		global $menu;
+
+		// The WooPayments plugin must not be active.
+		// When active, WooPayments will own the Payments menu item since it is the native Woo payments solution.
+		if ( $this->is_woopayments_active() ) {
+			return;
+		}
+
+		$menu_title = esc_html__( 'Payments', 'woocommerce' );
+		$menu_icon  = 'data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSI4NTIiIGhlaWdodD0iNjg0Ij48cGF0aCBmaWxsPSIjYTJhYWIyIiBkPSJNODIgODZ2NTEyaDY4NFY4NlptMCA1OThjLTQ4IDAtODQtMzgtODQtODZWODZDLTIgMzggMzQgMCA4MiAwaDY4NGM0OCAwIDg0IDM4IDg0IDg2djUxMmMwIDQ4LTM2IDg2LTg0IDg2em0zODQtNTU2djQ0aDg2djg0SDM4MnY0NGgxMjhjMjQgMCA0MiAxOCA0MiA0MnYxMjhjMCAyNC0xOCA0Mi00MiA0MmgtNDR2NDRoLTg0di00NGgtODZ2LTg0aDE3MHYtNDRIMzM4Yy0yNCAwLTQyLTE4LTQyLTQyVjIxNGMwLTI0IDE4LTQyIDQyLTQyaDQ0di00NHoiLz48L3N2Zz4=';
```
Comment (oaratovskyi) [inline plugins/woocommerce/src/Internal/Admin/Settings/PaymentsController.php:72]: Is settings image like that is common way in WC Core?

File: plugins/woocommerce/src/Internal/Admin/Settings/PaymentsController.php [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+			$menu_title,
+			'manage_woocommerce', // Capability required to see the menu item.
+			$menu_path,
+			null,
+			$menu_icon,
+			56, // Position after WooCommerce Product menu item.
+		);
+
+		// If the store doesn't have any enabled gateways or providers need action, add a notice badge to the Payments menu item.
+		if ( ! $this->store_has_enabled_gateways() || $this->store_has_providers_with_action() ) {
```
Comment (oaratovskyi) [inline plugins/woocommerce/src/Internal/Admin/Settings/PaymentsController.php:87]: `$this->store_has_providers_with_action()`

I guess that's a good idea to show the badge if there's any action.

File: plugins/woocommerce/src/Internal/Admin/Settings/PaymentsController.php [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+			// In case of an error, just return false.
+			return false;
+		}
+
+		// Go through the providers and check if any of them need attention from the user.
+		foreach ( $providers as $provider ) {
+			// Handle payment gateways and offline payment methods that need setup.
+			if (
+				in_array( $provider['_type'], array( Payments::PROVIDER_TYPE_GATEWAY, Payments::PROVIDER_TYPE_OFFLINE_PM ), true ) &&
+				! empty( $provider['state']['enabled'] ) &&
```
Comment (oaratovskyi) [inline plugins/woocommerce/src/Internal/Admin/Settings/PaymentsController.php]: I think that if it needs setup, then it won't be enabled yet. at least that's what I noticed during my testing with Stripe, WooPayments. in this case, we'll show the badge even if it's enabled, but needs setup - I guess no harm in that.

**Review by oaratovskyi — APPROVED** [review]
Testing instrcutions are awesome and they check out 🚀 
Approving 🟢

File: plugins/woocommerce/src/Admin/Features/OnboardingTasks/Tasks/Payments.php (hunk 2 above)
Comment (vladolaru) [reply plugins/woocommerce/src/Admin/Features/OnboardingTasks/Tasks/Payments.php:118]: Yeah. That is how the tasks work.

File: plugins/woocommerce/src/Internal/Admin/Settings/PaymentsController.php (hunk 3 above)
Comment (vladolaru) [reply plugins/woocommerce/src/Internal/Admin/Settings/PaymentsController.php:72]: I don't really know. We've done it like that going back around two years.

File: plugins/woocommerce/src/Internal/Admin/Settings/PaymentsController.php (hunk 5 above)
Comment (vladolaru) [reply plugins/woocommerce/src/Internal/Admin/Settings/PaymentsController.php]: > I think that if it needs setup, then it won't be enabled yet.

This is very much at each PE's discretion. Yes, WooPayments and the Stripe official PE work like that, but not all of them.

> in this case, we'll show the badge even if it's enabled, but needs setup

Quite the contrary. If it is NOT enabled but needs setup, we won't show the badge (the conditions have AND between them). But you do raise a good issue. We should not look at `enabled`, just at `needs_setup`. Changing it.

**Pull request merged**